[17.0][mig] purchase_analytic

aisopuro opened the pull request:
Replaces (resurrects?) #626

Commit:
module to manage analytic on purchase order

Commit:
[10.0][MIG] port purchase_analytic to V10.0

oca-travis committed:
[UPD] Update purchase_analytic.pot

OCA-git-bot committed:
[ADD] icon.png

MiquelRForgeFlow committed:
[MIG] purchase_analytic: Migration to 12.0

oca-travis committed:
[UPD] Update purchase_analytic.pot

OCA-git-bot committed:
[UPD] README.rst

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-analytic-12.0/account-analytic-12.0-purchase_analytic
Translate-URL: https://translation.odoo-community.org/projects/account-analytic-12-0/account-analytic-12-0-purchase_analytic/

ps-tubtim committed:
[IMP] purchase_analytic: black, isort

ps-tubtim committed:
[MIG] purchase_analytic: Migration to 13.0

oca-travis committed:
[UPD] Update purchase_analytic.pot

eduardgm committed:
Added translation using Weblate (Spanish)

alan196 committed:
[IMP] purchase_analytic: black, isort, prettier

alan196 committed:
[MIG] purchase_analytic: Migration to 15.0

Commit:
[UPD] Update purchase_analytic.pot

OCA-git-bot committed:
[UPD] README.rst

LoisRForgeFlow committed:
[IMP] purchase_analytic: adapt to newest framework.

Some of the "tricks" done in this module are no longer needed
and can be easily implemented with newest framework features:

* No need for an auxiliar `project_id2` field. User can set
  an analytic account with no lines and it is respected.
* Simplify onchange. Now update analytic line on the go (no
  need to save) which is a better UX because avoid unexpected
  changes on save.

Also re-label the field `project_id` to "Analytic Account" to
align with the typical label in newer versions of Odoo.
Tha label "Contract / Analytic" was last used in v8 (
https://github.com/odoo/odoo/blob/8.0/addons/sale/sale.py#L217).

Commit:
[UPD] Update purchase_analytic.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
purchase_analytic 15.0.2.0.0

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-analytic-15.0/account-analytic-15.0-purchase_analytic
Translate-URL: https://translation.odoo-community.org/projects/account-analytic-15-0/account-analytic-15-0-purchase_analytic/

AungKoKoLin1997 committed:
[MIG] purchase_analytic: Migration to 16.0

AungKoKoLin1997 committed:
analytic_distribution readonly states

Commit:
[UPD] Update purchase_analytic.pot

OCA-git-bot committed:
[UPD] README.rst

badbole committed:
Added translation using Weblate (Croatian)

badbole committed:
Translated using Weblate (Croatian)

Currently translated at 100.0% (3 of 3 strings)

Translation: account-analytic-16.0/account-analytic-16.0-purchase_analytic
Translate-URL: https://translation.odoo-community.org/projects/account-analytic-16-0/account-analytic-16-0-purchase_analytic/hr/

Gelo-fl committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (3 of 3 strings)

Translation: account-analytic-16.0/account-analytic-16.0-purchase_analytic
Translate-URL: https://translation.odoo-community.org/projects/account-analytic-16-0/account-analytic-16-0-purchase_analytic/es/

sanrav committed:
Translated using Weblate (French)

Currently translated at 100.0% (3 of 3 strings)

Translation: account-analytic-16.0/account-analytic-16.0-purchase_analytic
Translate-URL: https://translation.odoo-community.org/projects/account-analytic-16-0/account-analytic-16-0-purchase_analytic/fr/

Commit:
[UPD] Update purchase_analytic.pot

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-analytic-16.0/account-analytic-16.0-purchase_analytic
Translate-URL: https://translation.odoo-community.org/projects/account-analytic-16-0/account-analytic-16-0-purchase_analytic/

sanrav committed:
Translated using Weblate (French)

Currently translated at 100.0% (4 of 4 strings)

Translation: account-analytic-16.0/account-analytic-16.0-purchase_analytic
Translate-URL: https://translation.odoo-community.org/projects/account-analytic-16-0/account-analytic-16-0-purchase_analytic/fr/

sanrav committed:
Added translation using Weblate (French (France) (fr_FR))

sanrav committed:
Translated using Weblate (French (France) (fr_FR))

Currently translated at 100.0% (4 of 4 strings)

Translation: account-analytic-16.0/account-analytic-16.0-purchase_analytic
Translate-URL: https://translation.odoo-community.org/projects/account-analytic-16-0/account-analytic-16-0-purchase_analytic/fr_FR/

AungKoKoLin1997 committed:
[IMP] purchase_analytic: add default value in order line

OCA-git-bot committed:
purchase_analytic 16.0.2.0.0

adrianojprado committed:
Added translation using Weblate (Portuguese (Brazil))

adrianojprado committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (4 of 4 strings)

Translation: account-analytic-16.0/account-analytic-16.0-purchase_analytic
Translate-URL: https://translation.odoo-community.org/projects/account-analytic-16-0/account-analytic-16-0-purchase_analytic/pt_BR/

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (4 of 4 strings)

Translation: account-analytic-16.0/account-analytic-16.0-purchase_analytic
Translate-URL: https://translation.odoo-community.org/projects/account-analytic-16-0/account-analytic-16-0-purchase_analytic/es/

OCA-git-bot committed:
[UPD] README.rst

AungKoKoLin1997 committed:
[FIX] purchase_analytic: Access Denied by ACLs

OCA-git-bot committed:
[BOT] post-merge updates

sbidoul committed:
[TRN] remove fr_FR.po

Co-authored-by: Alexis de Lattre <alexis@via.ecp.fr>

Wodran14 committed:
[IMP] purchase_analytic: pre-commit stuff

Wodran14 committed:
[MIG] purchase_analytic: Migration to 17.0

aisopuro commented:
Trying to re-open this, since the original PR was closed due to inactivity. I only did some pre-commit fixes and squashed them into @Wodran14 's original pre-commit commit.

aisopuro commented:
The coverage appears to be failing due to there not being tests for "empty cases", e.g. computing the value correctly if there are no lines on an order.

Could the maintainers give an opinion, is it required to meet those code coverage goals on a migration PR?

StefanRijnhart commented:
/ocabot migration purchase_analytic

CLaurelB commented:
Hello,

Functional test: LGTM 👍 

Regarding the coverage: IMHO, there could be a test to remove the analytic distribution on the PO and then assert that the analytic distributions of the lines, previously set, remain intact. Additionally, there could be a test to set the same analytic distribution on multiple lines and verify that the analytic distribution of the PO is correctly set.

Best regards.

CLaurelB committed:
[IMP] purchase_analytic: improve tests to increase coverage

Add cases for setting the analytic distribution on a purchase order
without lines and unsetting the analytic distribution on a purchase
order with lines.

CLaurelB commented:
Hello @aisopuro,

Maybe this could help: https://github.com/avoinsystems/account-analytic/pull/2

Best regards.

aisopuro commented:
OK, thank you @CLaurelB for the coverage expansion, that commit is now included here. All checks are green so this seems ready to review by maintainers.

luisg123v commented on purchase_analytic/views/purchase_views.xml:
```diff
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="utf-8" ?>
+<?xml version="1.0" ?>
```
Why this? I'd avoid this, AFAIK it's preferred to provide the encoding explicitly.

luisg123v commented on purchase_analytic/i18n/purchase_analytic.pot:
```diff
@@ -0,0 +1,34 @@
+# Translation of Odoo Server.
```
There are missing commits present in 16.0, mainly related to translations, e.g. de8e27fa84267343190adc2874042efe9593cddf. I'd suggest retrieving the history again and then cherrypicking the migration commits.

luisg123v commented on purchase_analytic/models/purchase.py:
```diff
@@ -0,0 +1,44 @@
+
+    def _inverse_analytic_distribution(self):
+        """
+        When set analytic_distribution set analytic distribution on all order lines
```
"when set**ting**

Same in the below docstring.

luisg123v commented on purchase_analytic/tests/test_purchase_analytic.py:
```diff
@@ -17,65 +18,74 @@ def setUp(self):
+                    "name": self.product_id.name,
+                    "product_id": self.product_id.id,
+                    "product_qty": 1.0,
+                    "product_uom": self.uom_id.id,
```
I'd suggest making these params configurable and provide default values, so it may be called without params but also being able to provide some of them.

luisg123v commented on purchase_analytic/tests/test_purchase_analytic.py:
```diff
@@ -17,65 +18,74 @@ def setUp(self):
-    def test_analytic_distribution(self):
+    def add_new_order_line(self):
+        return [
+            Command.create(
```
Why don't you pass the order as param instead?

luisg123v commented on purchase_analytic/tests/test_purchase_analytic.py:
```diff
@@ -12,9 +12,7 @@ def setUp(self):
-        analytic_plan = self.env["account.analytic.plan"].create(
-            {"name": "Plan Test", "company_id": False}
-        )
+        analytic_plan = self.env["account.analytic.plan"].create({"name": "Plan Test"})
```
Why this? Could you document it in commit description?

luisg123v commented on purchase_analytic/README.rst:
```diff
@@ -0,0 +1,125 @@
+Contributors
+------------
+
+-  Laetitia Gangloff <laetitia.gangloff@acsone.eu>
```
Isn't fixing double spaces part of the changes performed by precommit?

I mean:

```suggestion
- Laetitia Gangloff <laetitia.gangloff@acsone.eu>
```

aisopuro commented:
Heads up, I may not have time anytime soon to fix the issues brought up by the review. So if someone wants to push this through, feel free to re-steal the PR from me.

StefanRijnhart commented:
Closing in favour of https://github.com/OCA/account-analytic/pull/731. Thank you for your work!
